src/util-flow-rate.c

```diff
@@ -0,0 +1,626 @@
+void FlowRateRegisterFlowStorage(void)
+{
+    ConfNode *root = ConfGetNode("flow");
+    if (root == NULL)
+        return;
+
+    bool track_flow = false;
+    track_flow = ConfNodeLookupChild(root, "rate-tracking") != NULL ? true : false;
+    if (!track_flow)
+        return;
+
+    ConfNode *node = ConfGetNode("flow.rate-tracking");
+    const char *val = ConfNodeLookupChildValue(node, "bytes");
+    if (val == NULL) {
+        FatalError("No value for flow tracking bytes");
+    }
+    uint64_t bytes = 0;
+    if (ParseSizeStringU64(val, &bytes) < 0) {
+        FatalError("Invalid value for flow tracking bytes");
+    }
+    flow_rate_config.bytes = bytes;
+
+    val = ConfNodeLookupChildValue(node, "interval");
+    if (val == NULL) {
+        FatalError("No value for flow tracking interval");
+    }
+    SCTime_t interval = SCTIME_INITIALIZER;
+    uint16_t secs = 0;
+    if ((StringParseUint16(&secs, 10, 0, val) < 0) || (secs == 0)) {
+        FatalError("Invalid value for flow tracking interval");
+    }
+    flow_rate_config.interval = SCTIME_ADD_SECS(interval, secs);
+
+    g_flowrate_storage_id = FlowStorageRegister(
+            "flowrate", sizeof(void *), NULL, (void (*)(void *))FlowRateStoreFree);
+}
+
+bool FlowRateStorageEnabled(void)
+{
+    return (g_flowrate_storage_id.id != -1);
+}
+
+FlowRateStore *FlowRateStoreInit(void)
+{
+    FlowRateStore *frs = NULL;
+    size_t total_memuse = 0;
+    size_t expected_memuse = (2 * flow_rate_config.interval.secs * sizeof(uint64_t)) + sizeof(*frs);
+
+    if (!FLOW_CHECK_MEMCAP(expected_memuse)) {
+        return NULL;
+    }
+    frs = SCCalloc(1, sizeof(*frs));
+    if (unlikely(frs == NULL)) {
+        SCLogError("Unable to allocate FlowRateStore memory");
+        return NULL;
+    }
+
+    total_memuse += sizeof(*frs);
+    for (int i = 0; i < 2; i++) {
+        frs->dir[i].size = flow_rate_config.interval.secs;
+        frs->dir[i].buf = SCCalloc(frs->dir[i].size, sizeof(uint64_t));
+        if (unlikely(frs->dir[i].buf == NULL)) {
+            FlowRateStoreFree(frs);
+            return NULL;
+        }
+        frs->dir[i].sum = 0;
+        frs->dir[i].last_idx = 0;
+        frs->dir[i].start_ts = SCTIME_INITIALIZER;
+        frs->dir[i].last_ts = SCTIME_INITIALIZER;
+        total_memuse += (frs->dir[i].size * sizeof(uint64_t));
+    }
+    DEBUG_VALIDATE_BUG_ON(total_memuse != expected_memuse);
+    (void)SC_ATOMIC_ADD(flow_memuse, total_memuse);
+
+    return frs;
+}
+
+FlowStorageId FlowRateGetStorageID(void)
+{
+    return g_flowrate_storage_id;
+}
+
+static inline void FlowRateClearSumTill(
+        FlowRateStore *frs, uint16_t start, uint16_t end, int direction)
+{
+    for (uint16_t i = start; i <= end; i++) {
+        uint64_t byte_count_at_i = frs->dir[direction].buf[i];
+        frs->dir[direction].buf[i] = 0;
+        frs->dir[direction].sum -= byte_count_at_i;
+    }
+}
+
+void FlowRateStoreUpdate(FlowRateStore *frs, SCTime_t p_ts, uint32_t pkt_len, int direction)
```

Comment:
style: I think this function is a bit long, can the 2 main branches be put into small (inline) helper funcs?